jpa/WFLY-11233_EntityManagerCache.adoc

```diff
@@ -0,0 +1,86 @@
+
+    The container must throw the TransactionRequiredException if a transaction-scoped persistence context is used and the EntityManager persist, remove, merge, or refresh method is invoked when no transaction is active.
+
+    For stateful session beans with extended persistence contexts:
+
+        The container creates an entity manager by calling EntityManagerFactory.createEntityManager when a stateful session bean is created that declares a dependency on an entity manager with PersistenceContextType.EXTENDED. (See section 7.6.3).
+        The container closes the entity manager by calling EntityManager.close after the stateful session bean and all other stateful session beans that have inherited the same persistence context as the entity manager have been removed.
+        When a business method of the stateful session bean is invoked, if the stateful session bean uses container managed transaction demarcation, and the entity manager is not already associated with the current JTA transaction, the container associates the entity manager with the current JTA transaction and, if the persistence context is of type SynchronizationType.SYNCHRONIZED, the container calls EntityManager.joinTransaction. If there is a different persistence context already associated with the JTA transaction, the container throws the EJBException.
+        When a business method of the stateful session bean is invoked, if the stateful session bean uses bean managed transaction demarcation and a UserTransaction is begun within the method, the container associates the persistence context with the JTA transaction and, if the persistence context is of type SynchronizationType.SYNCHRONIZED, the container calls EntityManager.joinTransaction.
+
+    The container must throw the IllegalStateException if the application calls EntityManager.close on a container-managed entity manager.
+
+    [87] It is not required that these contracts be used when a third-party persistence provider is not used: the container might use these same APIs or its might use its own internal APIs.
+
+    [88] The container may choose to pool EntityManagers: it instead of creating and closing in each case, it may acquire one from its pool and call clear() on it.
+
+One challenge with introducing an entity manager (per EntityManagerFactory) cache, is knowing when to close EntityManagers after certain errors have occurred on the EntityManager. To deal with this robustness issue, we will discard entity managers that are marked close when retrieving them from the cache. To ensure that we always put the cached EM back in cache, EntityManagerCache use is only done within JTA transactions, since we have control of the release (closing) of the EntityManager and only with SynchronizationType.SYNCHRONIZED (never with SynchronizationType.UNSYNCHRONIZED) persistence contexts.
```

Comment:
This concerns me a bit. This could leave `EntityManager`'s in the cache if one is not retrieved.

Reply:
I agree that caching EntityManager instances that may hold some state past the [EntityManager#clear](https://docs.oracle.com/javaee/7/api/javax/persistence/EntityManager.html#clear--) call (we call this when caching each entity manager).  

An example of an intermittent test failure that [we are seeing now is here](https://gist.github.com/scottmarlow/788a654b9dc53d078093ce64df9d4f26).  We will have to figure out this failure before this feature is complete.  We should review what the hibernate.connection.release_mode setting is, to see if that could be wrong.

Reply:
Also seeing another variation of error https://gist.github.com/scottmarlow/9edbe9131fee2a41b3267c78e06f2d0e

Reply:
I have yet to prove it but after [WFLY-11360](https://issues.jboss.org/browse/WFLY-11360) is solved, I will test that always running the JPA container Synchronization#afterCompletion call, after the Hibernate ORM Synchronization@afterCompletion call, will resolve the above errors.  I have a local coding change for this that will ensure that we don't put the EntityManager back in the cache, until after Hibernate is done making calls on it (which also includes protecting against the Transaction Manager REAPER thread cancelling transactions in the background).